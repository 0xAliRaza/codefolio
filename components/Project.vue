<template>
  <div class="project">
    <div class="project__image">
      <img
        :src="isValidUrl(image) ? image : require(`~/assets/images/${image}`)"
        alt="test image"
        class="img-fluid"
      />
      <div v-if="technologies" class="project__technologies">
        <span
          v-for="(tech, i) in technologies"
          :key="tech + i"
          class="project__technology"
          >{{ tech }}</span
        >
      </div>
    </div>
    <div class="project__details">
      <div class="project__title">
        <h3>{{ title }}</h3>
      </div>
      <div class="project__description">
        <p>{{ description }}</p>
      </div>
      <div v-if="links" class="project__links">
        <Button
          v-if="links.liveDemo"
          class=""
          :link="links.liveDemo"
          small
          primary
          title="Live Demo"
          >Demo
          <Icon name="globe" class="ms-1" />
        </Button>
        <Button
          v-if="links.sourceCode"
          class=""
          :link="links.sourceCode"
          small
          github
          title="Source Code"
          >Source
          <Icon name="github" class="ms-1" />
        </Button>
        <Button
          v-if="links.videoDemo"
          class=""
          :link="links.videoDemo"
          small
          youtube
          title="Video Demo"
          >Video
          <Icon name="youtubeplay" class="ms-1" />
        </Button>
      </div>
    </div>
  </div>
</template>
<script lang="ts">
import Vue from 'vue'
export default Vue.extend({
  props: {
    title: { required: true, type: String },
    description: { required: true, type: String },
    image: { required: true, type: String },
    links: { default: null, type: Object },
    technologies: { default: null, type: Array },
  },
  methods: {
    isValidUrl: function (str: string) {
      let url

      try {
        url = new URL(str)
      } catch (_) {
        return false
      }

      return url.protocol === 'http:' || url.protocol === 'https:'
    },
  },
})
</script>
<style lang="scss" scoped>
@import '@/scss/abstracts';
.project {
  background: $light;
  border: 1px solid $gray-300;
  overflow: hidden;
  max-width: pxToRem(480);
  margin-right: auto;
  margin-left: auto;
  border-radius: pxToRem(6px);
  transition: border 0.2s ease-in-out, box-shadow 0.2s ease-in-out,
    transform 0.2s ease-in-out;
  @include hocus {
    transform: translateY(-2px);
    border: 1px solid $gray-400;
    box-shadow: 0 2px pxToRem(16) rgba($gray-800, 0.3);
  }
  &__details {
    padding: pxToRem(16);
  }
  &__title {
    h3 {
      font-weight: 600;
      color: $gray-800;
      font-size: pxToRem(26);
    }
  }
  &__description {
    p {
      font-size: pxToRem(15);
      line-height: 1.5em;
      margin: 0;
      color: $gray-700;
    }
  }

  &__links {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
    padding: pxToRem(22) 0;
    gap: pxToRem(6);
  }

  &__technologies {
    padding: 0 pxToRem(16);
    margin: pxToRem(8) 0;
    display: flex;
    align-items: center;
    flex-wrap: wrap;
    gap: pxToRem(4);
  }
  &__technology {
    line-height: 1.3;
    text-transform: uppercase;
    background: $gray-200;
    color: $gray-600;
    border-radius: pxToRem(4);
    padding: pxToRem(5);
    font-size: pxToRem(12);
    font-weight: 500;
  }
}
</style>
