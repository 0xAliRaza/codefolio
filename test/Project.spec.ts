import { shallowMount } from '@vue/test-utils'
import Project from '@/components/Project.vue'

const dummyProps: any = {
  title: 'test_title_lkjsadfk;lkjl;jasdf',
  description: 'test_description_ljsdfljk128734',
  image:
    'https://images.unsplash.com/photo-1633000975009-b1ce39994b81?crop=entropy&cs=tinysrgb&fit=crop&fm=jpg&h=400&ixid=MnwxfDB8MXxyYW5kb218MHx8fHx8fHx8MTYzNDExMzE4Ng&ixlib=rb-1.2.1&q=80&w=800',
  links: {
    liveDemo: 'https://0xali.com',
    videoDemo: 'https://youtube.com/',
    sourceCode: 'https://github.com/',
  },
  technologies: [
    'HTML',
    'CSS',
    'JavaScript',
    'VueJS',
    'NuxtJS',
    'Docker',
    'SCSS',
    'Git',
  ],
}

const factory = (props = {}) => {
  return shallowMount(Project, {
    propsData: { ...props },
    stubs: ['Button', 'Icon'],
  })
}
describe('Project', () => {
  it('renders correctly', () => {
    const wrapper = factory(dummyProps)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders project title', () => {
    const wrapper = factory(dummyProps)
    expect(wrapper.text()).toContain(dummyProps.title)
  })

  it('renders project description', () => {
    const wrapper = factory(dummyProps)
    expect(wrapper.text()).toContain(dummyProps.description)
  })

  it('renders image tag', () => {
    const wrapper = factory(dummyProps)
    const img = wrapper.get(`img[src="${dummyProps.image}"]`)
    expect(img).toBeTruthy()
  })
  it('renders live demo button', () => {
    const wrapper = factory(dummyProps)
    const a = wrapper.get(`button-stub[link="${dummyProps.links.liveDemo}"]`)
    expect(a).toBeTruthy()
  })
  it('renders source code button', () => {
    const wrapper = factory(dummyProps)
    const a = wrapper.get(`button-stub[link="${dummyProps.links.sourceCode}"]`)
    expect(a).toBeTruthy()
  })
  it('renders video demo button', () => {
    const wrapper = factory(dummyProps)
    const a = wrapper.get(`button-stub[link="${dummyProps.links.videoDemo}"]`)
    expect(a).toBeTruthy()
  })
  it('renders all the technologies', () => {
    const wrapper = factory(dummyProps)
    const text = wrapper.text()
    dummyProps.technologies.forEach((tech: string) => {
      expect(text).toContain(tech)
    })
  })
})
