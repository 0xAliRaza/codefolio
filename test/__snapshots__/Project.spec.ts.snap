// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project renders correctly 1`] = `
<div class="project">
  <div class="project__image">
    <img src="https://images.unsplash.com/photo-1633000975009-b1ce39994b81?crop=entropy&amp;cs=tinysrgb&amp;fit=crop&amp;fm=jpg&amp;h=400&amp;ixid=MnwxfDB8MXxyYW5kb218MHx8fHx8fHx8MTYzNDExMzE4Ng&amp;ixlib=rb-1.2.1&amp;q=80&amp;w=800"
         alt="test image"
         class="img-fluid"
    >
    <div class="project__technologies">
      <span class="project__technology">
        HTML
      </span>
      <span class="project__technology">
        CSS
      </span>
      <span class="project__technology">
        JavaScript
      </span>
      <span class="project__technology">
        VueJS
      </span>
      <span class="project__technology">
        NuxtJS
      </span>
      <span class="project__technology">
        Docker
      </span>
      <span class="project__technology">
        SCSS
      </span>
      <span class="project__technology">
        Git
      </span>
    </div>
  </div>
  <div class="project__details">
    <div class="project__title">
      <h3>
        test_title_lkjsadfk;lkjl;jasdf
      </h3>
    </div>
    <div class="project__description">
      <p>
        test_description_ljsdfljk128734
      </p>
    </div>
    <div class="project__links">
      <button-stub link="https://0xali.com"
                   small
                   primary
                   title="Live Demo"
      >
        Demo
        <icon-stub name="globe"
                   class="ms-1"
        >
        </icon-stub>
      </button-stub>
      <button-stub link="https://github.com/"
                   small
                   github
                   title="Source Code"
      >
        Source
        <icon-stub name="github"
                   class="ms-1"
        >
        </icon-stub>
      </button-stub>
      <button-stub link="https://youtube.com/"
                   small
                   youtube
                   title="Video Demo"
      >
        Video
        <icon-stub name="youtubeplay"
                   class="ms-1"
        >
        </icon-stub>
      </button-stub>
    </div>
  </div>
</div>
`;
